File: 2 term/PB/PB Lab2.2/PB Lab2.2 cs/PB Lab2.2 cs/IOManager.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace PB_Lab2._2_cs
{
    public class IOManager
    {
        public void GetClients(string fileName)
        {
            FileManager fileManager = new FileManager();
            string name;
            string? comingTime;
            string? leavingTime;
            List<Client> clients = fileManager.ReadAllClients(fileName);
            Console.WriteLine("To input a new client, press a, else press Enter");
            string? input = Console.ReadLine();
            while(input == "a")
            {
                Console.Write("Enter clients name: ");
                name = Console.ReadLine() ?? " ";
                Console.Write("Enter coming time: ");
                comingTime = Console.ReadLine();
                Console.Write("Enter leaving time: ");
                leavingTime = Console.ReadLine();
                if (CheckInput(comingTime, leavingTime, clients))
                {
                    Client client = new Client();
                    client.Name = name;
                    client.ComingTime = TimeOnly.Parse(comingTime);
                    client.LeavingTime = TimeOnly.Parse(leavingTime);
                    clients.Add(client);
                    fileManager.WriteClient(client, fileName);
                    Console.WriteLine("Successfully added");
                }
                else
                {
                    Console.WriteLine("Either time format is wrong or timespan isn't free, client not added");
                }
                Console.WriteLine("To input a new client, press a, else press Enter");
                input = Console.ReadLine();
            }
        }
        public bool CheckInput(string? comingTime, string? leavingTime, List<Client> clients)
        {
            TimeOnly firstTime;
            TimeOnly secondTime;
            if (TimeOnly.TryParse(comingTime, out firstTime) && TimeOnly.TryParse(leavingTime, out secondTime))
            {
                foreach (Client client in clients)
                {
                    if (firstTime < client.LeavingTime && firstTime > client.ComingTime ||
                        secondTime < client.LeavingTime && secondTime > client.ComingTime ||
                        firstTime < client.ComingTime && secondTime > client.LeavingTime ||
                        firstTime > secondTime ||
                        firstTime < Client.WorkStartTime ||
                        secondTime > Client.WorkEndTime)
                    {
                        return false;
                    } 
                }
            }
            else
            {
                return false;
            }
            return true;
        }
    }
}

File: 2 term/PB/PB Lab2.2/PB lab 2.2 py/io_manager.py
from Client import *
from file_manager import *


def input_clients(file_name, writing_option):
    old_clients = []
    new_clients = []
    if writing_option == 'ab':
        try:
            for client in read_all_clients(file_name):
                for i in range(len(client)):
                    old_clients.append(client[i])
        except FileNotFoundError:
            pass
    character = input("If you want to add client, press a, otherwise press Enter: ")
    while character == 'a':
        name = input("Enter client's name: ")
        coming_time = input("Enter coming time(hh:mm): ")
        leaving_time = input("Enter leaving time(hh:mm): ")
        client = Client(name, coming_time, leaving_time)
        if check_client(coming_time, leaving_time, old_clients) and check_client(coming_time, leaving_time, new_clients):
            new_clients.append(client)
            print("Added successfully")
        else:
            print("Wrong time entered, client wasn't added")
        character = input("If you want to add client, press a, otherwise press Enter: ")
    return new_clients


def output_clients(file_name):
    clients = []
    for client in read_all_clients(file_name):
        for i in range(len(client)):
            clients.append(client[i])
    clients.sort(key=lambda x: get_minutes_from_time(x.coming_time))
    for client in clients:
        print(f"{client.name}: {client.coming_time} - {client.leaving_time}")


def ask_writing_option():
    writing_option = input("If you want to append existing file, press a, otherwise press Enter: ")
    if writing_option == 'a':
        return 'ab'
    else:
        return 'wb'


def print_free_time(file_name):
    first_half, second_half = get_free_time(file_name)
    print(f"Number of free periods in the first half of the day: {first_half}")
    print(f"Number of free periods in the second half of the day: {second_half}")


def check_client(coming_time, leaving_time, clients):
    coming_minutes = get_minutes_from_time(coming_time)
    leaving_minutes = get_minutes_from_time(leaving_time)
    if (coming_minutes > leaving_minutes or
        coming_minutes < 9 * 60 or
        leaving_minutes > 17 * 60):
        return False
    for client in clients:
        current_coming_minutes = get_minutes_from_time(client.coming_time)
        current_leaving_minutes = get_minutes_from_time(client.leaving_time)
        if (coming_minutes < current_leaving_minutes < leaving_minutes or
            coming_minutes < current_coming_minutes < leaving_minutes or
            coming_minutes < current_coming_minutes and current_leaving_minutes < leaving_minutes
            ):
            return False
    return True


def get_minutes_from_time(time):
    return int(time.split(":")[0]) * 60 + int(time.split(":")[1])

